test/unit/org/apache/cassandra/tools/ToolRunner.java

```diff
@@ -273,228 +109,466 @@ public void checkPermission(Permission perm, Object context)
+         * stdErr
+         * 
+         * @param regExpCleaners
+         *            List of regExps to remove from stdErr
+         * @return The stdErr with all excludes removed
+         */
+        public String getCleanedStderr(List<String> regExpCleaners)
+        {
+            String sanitizedStderr = getStderr();
+            for (String regExp : regExpCleaners)
+                sanitizedStderr = sanitizedStderr.replaceAll(regExp, "");
+            return sanitizedStderr;
+        }
+
+        /**
+         * Returns stdErr after removing any potential JVM env info output. Uses default list of excludes
+         * 
+         * {@link #getCleanedStderr(List)}
+         */
+        public String getCleanedStderr()
+        {
+            return getCleanedStderr(DEFAULT_CLEANERS);
+        }
+        
+        public void assertOnCleanExit()
+        {
+            assertOnExitCode();
+            assertCleanStdErr();
+        }
+    }
+
+    public interface ObservableTool extends AutoCloseable
+    {
+        String getPartialStdout();
+
+        String getPartialStderr();
+
+        boolean isDone();
+
+        ToolResult waitComplete();
+
+        @Override
+        void close();
     }
 
-    public static class Runners
+    private static final class FailedObservableTool implements ObservableTool
     {
-        public static ToolRunner invokeNodetool(String... args)
+        private final List<String> args;
+        private final IOException error;
+
+        private FailedObservableTool(IOException error, List<String> args)
+        {
+            this.args = args;
+            this.error = error;
+        }
+
+        @Override
+        public String getPartialStdout()
+        {
+            return "";
+        }
+
+        @Override
+        public String getPartialStderr()
+        {
+            return error.getMessage();
+        }
+
+        @Override
+        public boolean isDone()
         {
-            return invokeNodetool(Arrays.asList(args));
+            return true;
         }
 
-        public static ToolRunner invokeNodetool(List<String> args)
+        @Override
+        public ToolResult waitComplete()
         {
-            return invokeTool(buildNodetoolArgs(args), true);
+            return new ToolResult(args, -1, getPartialStdout(), getPartialStderr(), error);
         }
 
-        private static List<String> buildNodetoolArgs(List<String> args)
+        @Override
+        public void close()
         {
-            return CQLTester.buildNodetoolArgs(args);
+
         }
+    }
 
-        public static ToolRunner invokeClassAsTool(String... args)
+    private static final class ForkedObservableTool implements ObservableTool
+    {
+        private final CompletableFuture<Void> onComplete = new CompletableFuture<>();
+        private final ByteArrayOutputStream err = new ByteArrayOutputStream();
```

Comment:
would need to load in IntelliJ but I believe this would be an IntelliJ warning, so I think we should add `@SupressWarning("resources")`

Reply:
Mine says they are unneeded, but I added them as I know you get those flagged :shrug: